resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala

```diff
@@ -110,30 +109,29 @@ private[spark] class Client(
     for (nextStep <- submissionSteps) {
       currentDriverSpec = nextStep.configureDriver(currentDriverSpec)
     }
-
-    val resolvedDriverJavaOpts = currentDriverSpec
-      .driverSparkConf
-      // Remove this as the options are instead extracted and set individually below using
-      // environment variables with prefix SPARK_JAVA_OPT_.
-      .remove(org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
-      .getAll
-      .map {
-        case (confKey, confValue) => s"-D$confKey=$confValue"
-      } ++ driverJavaOptions.map(Utils.splitCommandString).getOrElse(Seq.empty)
-    val driverJavaOptsEnvs: Seq[EnvVar] = resolvedDriverJavaOpts.zipWithIndex.map {
-      case (option, index) =>
-        new EnvVarBuilder()
-          .withName(s"$ENV_JAVA_OPT_PREFIX$index")
-          .withValue(option)
-          .build()
-    }
-
+    val configMapName = s"$kubernetesResourceNamePrefix-driver-conf-map"
+    val configMap = buildConfigMap(configMapName, currentDriverSpec.driverSparkConf)
+    // The include of the ENV_VAR for "SPARK_CONF_DIR" is to allow for the
+    // Spark command builder to pickup on the Java Options present in the ConfigMap
     val resolvedDriverContainer = new ContainerBuilder(currentDriverSpec.driverContainer)
-      .addAllToEnv(driverJavaOptsEnvs.asJava)
+      .addNewEnv()
+        .withName(SPARK_CONF_DIR_ENV)
+        .withValue(SPARK_CONF_PATH)
```

Comment:
I have one concern on using a different path for `SPARK_CONF_DIR` than `/opt/spark/conf` which is the one under the installation path. Because the environment variable `SPARK_CONF_DIR` is set, people who want to customize the image, e.g., using a custom `spark-env.sh` has to know to put the file into this custom path instead of `/opt/spark/conf`. Why don't we use `/opt/spark/conf` instead?

Reply:
I am flexible to that, anyone else have thoughts?

Reply:
I'm unsure of what the behavior is if we try to mount to a directory that already exists on the image itself.

Reply:
Then I would suggest removing `COPY conf /opt/spark/conf` from the Dockerfile as it is no effect anyway if `SPARK_CONF_DIR` points to somewhere else. Then set `SPARK_CONF_DIR` to point to `/opt/spark/conf`.

Reply:
+1 for this.

Reply:
Do the executors require a SPARK_CONF_DIR directory to be defined as well?

Reply:
Yes.